ext/re2/re2.cc

```diff
@@ -1563,13 +1583,23 @@ static VALUE re2_set_initialize(int argc, VALUE *argv, VALUE self) {
 static VALUE re2_set_add(VALUE self, VALUE pattern) {
   StringValue(pattern);
   re2::StringPiece regex(RSTRING_PTR(pattern), RSTRING_LEN(pattern));
-  std::string err;
   re2_set *s;
   Data_Get_Struct(self, re2_set, s);
 
-  int index = s->set->Add(regex, &err);
+  /* To prevent the memory of the err string leaking when we call rb_raise,
+   * take a copy of it and let it go out of scope.
+   */
+  char msg[100];
+  int index;
+
+  {
+    std::string err;
+    index = s->set->Add(regex, &err);
+    strncpy(msg, err.c_str(), sizeof(msg));
```

Comment:
We should also guard against the error message being truncated (and therefore not null terminated) with something like:

```c++
strncpy(msg, err.c_str(), sizeof(msg) - 1);
msg[sizeof(msg) - 1] = '\0';
```

Reply:
Or better yet, use Ruby’s `strlcpy`: https://github.com/ruby/ruby/blob/ruby_2_6/missing/strlcpy.c